source/testcommon/README.md

```diff
@@ -1,9 +1,127 @@
-# Test Common Py
+# Testcommon utility tool
 
-This is common testing utilities for Python projects.
+This python packages includes tools and utility function for making testing easier.
+
+## Usage
+
+### Scenario testing
+
+To use testcommon for the scenario testing we need to set up three things.
+
+#### 1) Use the standard folder structure setup
+
+```plaintext
+├── scenario_name/
+    ├── when/
+        ├── input1.csv
+        └── ...
+    └── then/
+        ├── output1.csv
+        └── ...
+    └── test_output.py
+└── conftest.py
+```
+
+Input files are stored in the `/when` subfolder, while the expected results are stored in the `/then` subfolder.
+We prefix scenario names with `given_...` to enhance readability. As if we ready it line-by-line it makes sense.
+This is best shown with an example.
+
+```plaintext
+├──given_energy_scenario
+    ├── when/
+        ├── energy.csv
+        ├── energy_per_brp.csv
+        └── energy_per_es.csv
+    └── then/
+        └── energy_v1.csv
+```
+
+Notice, that there is not a one-to-one relation between the `/then` and `/when` files.
+This is because we may have to perform various transformations, such as aggregating or joining tables, to achieve the
+desired
+output. (i.e., `/then` file).
+
+#### 2) conftest.py
+
+The `conftest.py` file is a special pytest file where `fixtures` can be defined. `Fixtures` are small function that will
+run
+before the actual test runs (i.e. the `test_output.py`). As such we can use it for setting up the testing environment
+and
+therefore make the actual tests small and concise (i.e. limiting code duplication).
+
+We use the `conftest.py` to make a fixtures that performs 3 steps:
+
+1. Read `/when` file(s)
+1. Perform transformation(s)
+1. Construct a list `TestCases` containing `TestCase` object(s)
+
+The `TestCase` objects consist of two things.
+(a) A dataframe containing the transformed data.
+(b) A path for the expect csv file (i.e. the `/then` file).
+
+This means that the only thing we have left to do it to pass the `TestCases` object to the test files (`test_output.py`)
+to compare actual & expected from the `TestCase` objects.
+
+An example of how a `conftest.py` can look:
+
+```python
+@pytest.fixture(scope="module")
+def test_cases(spark: SparkSession, request: pytest.FixtureRequest):
+    # Setup
+    schema = StructType([
+        StructField("col1", StringType()),
+        StructField("col2", BooleanType()),
+        StructField("col2", IntegerType()),
+    ])
```

Comment:
I would put the schema outside the fixture so the fixture example is as succinct as possible.